### PR TITLE
Kulla operators

### PR DESCRIPTION
**Pull request by kudav**
- Many fixes for cleaning up the merge
- Work on coulomb log functions
- Added profile specification for S>1
- Minor improvements

### CODE REVIEW

**Review by lazersos — CHANGES_REQUESTED** [review]
Please fix all spacing back to what is in the repo.  It is impossible to tell what you've changed and what you haven't.

File: BEAMS3D/Sources/beams3d_distnorm.f90 [hunk 1]
```diff
@@ -39,94 +39,97 @@ SUBROUTINE beams3d_distnorm
...
+
+      ! Other two points
+      CALL beams3d_suv2rzp(s2,u1,p1,rt(3),zt(3),pt(3))
+      CALL beams3d_suv2rzp(s2,u2,p1,rt(4),zt(4),pt(4))
+
+      ! Assume parallel piped
+      area = 0.5 * abs( (rt(2) - rt(1)) *(zt(3) - zt(1)) - (rt(3) - rt(1)) *(zt(2) - zt(1))) + &
+         0.5 * abs( (rt(3) - rt(2)) *(zt(4) - zt(2)) - (rt(4) - rt(2)) *(zt(3) - zt(2)))
+
+      m = MIN(MAX(COUNT(phiaxis < p1),1),nphi-1)
```
Comment (lazersos) [inline BEAMS3D/Sources/beams3d_distnorm.f90]: Should be commented out.

File: BEAMS3D/Sources/beams3d_follow.f90 [hunk 2]
```diff
@@ -4,300 +4,301 @@
...
+   neut_lines(0, mystart:myend)   = .FALSE.
+   IF (lbeam) neut_lines(0, mystart:myend) = .TRUE.
 
-    ! Some helpers
-    fact_vsound = 1.5*sqrt(e_charge/plasma_mass)*therm_factor
-    fact_crit = SQRT(2*e_charge/plasma_mass)*(0.75*sqrt_pi*sqrt(plasma_mass/electron_mass))**(1.0/3.0) ! Wesson pg 226 5.4.9
-    !fact_kick = pi2*2*SQRT(pi*1E-7*plasma_mass)*E_kick*freq_kick !old
-    !fact_kick = 2*freq_kick*E_kick
+   ! Some helpers
+   fact_vsound = sqrt(3*e_charge/plasma_mass)*therm_factor
```
Comment (lazersos) [inline BEAMS3D/Sources/beams3d_follow.f90]: Why this change?

File: BEAMS3D/Sources/beams3d_init.f90 [hunk 3]
```diff
@@ -57,790 +58,788 @@ SUBROUTINE beams3d_init
...
+         k = COUNT(NI_AUX_S .ge. 0)
+         CALL EZspline_init(NI_spl_s(i),k,bcs1_s,ier)
+         IF (ier /=0) CALL handle_err(EZSPLINE_ERR,'beams3d_init9b',ier)
+         NI_spl_s(i)%isHermite   = 0
+         NI_spl_s(i)%x1          = NI_AUX_S(1:k)
+         CALL EZspline_setup(NI_spl_s(i),NI_AUX_F(i,1:k),ier,EXACT_DIM=.true.)
+         IF (ier /=0) CALL handle_err(EZSPLINE_ERR,'beams3d_init10b',ier)
+         IF (lverb .and. ANY(NI_AUX_F(i,:)>0)) WRITE(6,'(A,I1,A,F9.5,A,F9.5,A,I3,A,I2,A,F8.2)') '   Ni(',i,')= [', &
+            MINVAL(NI_AUX_F(i,1:k))*1E-20,',',MAXVAL(NI_AUX_F(i,1:k))*1E-20,'] E20 m^-3;  M: ',&
+            NINT(NI_AUX_M(i)/1.66053906660E-27),' amu;  Z: ',NI_AUX_Z(i), ';  S_MAX_NI = ',s_max_zeff
```
Comment (lazersos) [inline BEAMS3D/Sources/beams3d_init.f90:238]: Shouldn't this be s_max_ni?

File: BEAMS3D/Sources/beams3d_init.f90 [hunk 4]
```diff
@@ -57,790 +58,788 @@ SUBROUTINE beams3d_init
...
+      CALL EZspline_free(POT_spl,ier)
+   END IF
+   ! These are helpers for range
+   eps1 = (rmax-rmin)*small
+   eps2 = (phimax-phimin)*small
+   eps3 = (zmax-zmin)*small
+
+   ! Print Grid info to screen
+   IF (lverb) THEN
+      ! WRITE(6,'(A)')'----- Constructing Splines -----'
```
Comment (lazersos) [inline BEAMS3D/Sources/beams3d_init.f90]: Uncomment this.

File: BEAMS3D/Sources/beams3d_interface_mod.f90 [hunk 5]
```diff
@@ -173,132 +173,154 @@ SUBROUTINE beams3d_init_commandline
...
+               write(6, *) '     -boxsim:       Inject charged particles for box modeling'
+               write(6, *) '     -noverb:       Supress all screen output'
+               write(6, *) '     -help:         Output help message'
             end select
             i = i + 1
          END DO
          DEALLOCATE(args)
+
+         ! Handle particle restarting
+         IF (lrestart_particles .or. lrestart_grid) THEN
```
Comment (lazersos) [inline BEAMS3D/Sources/beams3d_interface_mod.f90]: Wouldn't you like the ability to use a restart grid with a new depo run?

File: BEAMS3D/Sources/beams3d_physics_mod.f90 [hunk 6]
```diff
@@ -7,2058 +7,2099 @@
...
+   !-----------------------------------------------------------------
+CONTAINS
+
+   !-----------------------------------------------------------------
+   !     Function:      coulomb_log_nrl19
+   !     Authors:       S. Lazerson (samuel.lazerson@ipp.mpg.de)
+   !     Date:          07/05/2023
+   !     Description:   Coulomb log as defined in NRL 2019
+   !-----------------------------------------------------------------
+   FUNCTION coulomb_log_nrl19(ne_in,te_in,vbeta_in,Zeff_in)  result(slow_par)
```
Comment (lazersos) [inline BEAMS3D/Sources/beams3d_physics_mod.f90:75]: What is the purpose of these changes? Also the documentation should be updated so that the function outputs are clear.

File: BEAMS3D/Sources/beams3d_physics_mod.f90 [hunk 7]
```diff
@@ -7,2058 +7,2099 @@
...
+         ! speed_cube = (speed*speed*speed)
+         ! CALL gauss_rand(1,zeta)  ! A random from a standard normal (1,1)
+         ! sigma = sqrt( ABS(2*e_charge*dt*(te_temp*speed_cube+ti_temp*vcrit_cube)*tau_spit_inv*inv_mymass/speed_cube) ) ! The standard deviation.
+         ! speed = speed+sigma*zeta
+
+
+         !------------------------------------------------------------
+         !  Pitch Angle Scattering
+         !------------------------------------------------------------
+         speed_cube = vc3_tauinv*slow_par(3)*dt/(speed*speed*speed) ! redefine as inverse
```
Comment (lazersos) [inline BEAMS3D/Sources/beams3d_physics_mod.f90:405]: ????  This is too many changes to a fundamental part of the code.  The operators should not be changed with out a published benchmark of tests.

File: BEAMS3D/Sources/beams3d_write.f90 [hunk 8]
```diff
@@ -5,390 +5,424 @@
...
+         IF (ier /= 0) CALL handle_err(HDF5_WRITE_ERR,'B_R',ier)
+         CALL write_var_hdf5(fid,'B_Z',nr,nphi,nz,ier,DBLVAR=B_Z,ATT='Vertical Trajectory Eq. (BZ)',ATT_NAME='description')
+         IF (ier /= 0) CALL handle_err(HDF5_WRITE_ERR,'B_Z',ier)
+         CALL write_var_hdf5(fid,'B_PHI',nr,nphi,nz,ier,DBLVAR=B_PHI,ATT='Toroidal Trajectory (BPHI)',ATT_NAME='description')
+         IF (ier /= 0) CALL handle_err(HDF5_WRITE_ERR,'B_PHI',ier)
+         CALL write_var_hdf5(fid,'S_ARR',nr,nphi,nz,ier,DBLVAR=S_ARR,ATT='Normalized Toroidal Flux',ATT_NAME='description')
+         IF (ier /= 0) CALL handle_err(HDF5_WRITE_ERR,'S_ARR',ier)
+         CALL write_var_hdf5(fid,'U_ARR',nr,nphi,nz,ier,DBLVAR=U_ARR,ATT='Equilibrium Poloidal Angle [rad]',ATT_NAME='description')
+         IF (ier /= 0) CALL handle_err(HDF5_WRITE_ERR,'U_ARR',ier)
+         CALL write_var_hdf5(fid,'X_ARR',nr,nphi,nz,ier,DBLVAR=X_ARR,ATT='Cartesian S/U grid',ATT_NAME='description')
```
Comment (lazersos) [inline BEAMS3D/Sources/beams3d_write.f90]: Not needed.

File: BEAMS3D/Sources/beams3d_write.f90 [hunk 9]
```diff
@@ -5,390 +5,424 @@
...
+         IF (ier /= 0) CALL handle_err(HDF5_WRITE_ERR,'X_ARR',ier)
+         CALL write_var_hdf5(fid,'Y_ARR',nr,nphi,nz,ier,DBLVAR=Y_ARR,ATT='Cartesian S/U grid',ATT_NAME='description')
```
Comment (lazersos) [inline BEAMS3D/Sources/beams3d_write.f90]: Not needed.

File: LIBSTELL/LIBSTELL.dep [hunk 10]
```diff
@@ -1310,9 +1310,6 @@ read_spec_mod.o : \
       read_wout_mod.o
       
 
-read_pies_mod.o : \
```
Comment (lazersos) [inline LIBSTELL/LIBSTELL.dep:1313]: This is absolutely needed.

File: LIBSTELL/ObjectList [hunk 11]
```diff
@@ -75,7 +75,6 @@ diagnostic_T.o \
 fftpack.o \
 vmec_utils.o \
 v3post_rfun.o \
-read_pies_mod.o \
```
Comment (lazersos) [inline LIBSTELL/ObjectList]: This is needed.

File: LIBSTELL/Sources/Modules/read_pies_mod.f90
Comment (lazersos) [inline LIBSTELL/Sources/Modules/read_pies_mod.f90:1]: Please undo.

**Review by lazersos — CHANGES_REQUESTED** [review]
I don't want new operators merged into develop until they've been tested and documented.  If you want to implement these changes then you at least need to show what the new operators do.  May be a good paper.

File: BEAMS3D/Sources/beams3d_init.f90 (hunk 3 above)
Comment (kudav) [reply BEAMS3D/Sources/beams3d_init.f90:238]: s_max_zeff is handled the same way as nzeff, which in our setup logic is used for NI_AUX_*, so  I didn't create an independent s_max_ni.

**Review by lazersos — COMMENTED** [review]
OK I need to run benchmarks on this branch to verify but it looks pretty good now.

**Review by lazersos — APPROVED** [review]
OK Looks good.
